### PR TITLE
Add a config option for the number of moduleslots in a chassipipe

### PR DESCRIPTION
**Pull request by Noradly**
I've made this pr because AE2 is IMHO far more powerful/cheaper and easier to craft while being less cool resulting in it being used more than it should(since this mods is far cooler).

Currently working
- Configurable number of slots in a chassipipe
- GUI creation at runtime (no idea if it's done 'correctly', but it's functional)
- The config settings (the default is the same as the current pipes)
- The modules & module options work in all the slots
- Reloading all textures

Not working
- Nothing?

### CODE REVIEW

**Comment by davboecki** [comment]
Have you looked at how our other GUIs are created?
We already have a system to dynamically create GUIs without any preconfigured texture. 
Other than that please take a look at our style for coding and especially for naming things.

**Comment by Noradly** [comment]
I've removed my runtime texture creator and `drawGuiContainerBackgroundLayer` now creates the texture at runtime, it should all still work right.

File: common/logisticspipes/gui/GuiChassiPipe.java [hunk 1]
```diff
@@ -209,9 +178,15 @@ private String getModuleName(int slot) {
 
 	@Override
 	protected void drawGuiContainerBackgroundLayer(float f, int x, int y) {
-		GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-		mc.renderEngine.bindTexture(_chassiPipe.getChassiGUITexture());
-		drawTexturedModalRect(guiLeft, guiTop, 0, 0, xSize, ySize);
+		GuiGraphics.drawGuiBackGround(mc, guiLeft, guiTop, right, bottom, zLevel, true);
+
+		GL11.glTranslated(guiLeft, guiTop, 0);
```
Comment (davboecki) [inline common/logisticspipes/gui/GuiChassiPipe.java]: Please don't use GL11 methods here, just put "guiLeft + 17" instead of 17 below, and so on...

**Comment by davboecki** [comment]
And please look at the final diff and clean up all the changes that are not necessary, especially the ones where files will not change otherwise.

**Comment by Noradly** [comment]
Should have changed al the things you asked for.

File: common/logisticspipes/gui/GuiChassiPipe.java [hunk 2]
```diff
@@ -209,9 +173,13 @@ private String getModuleName(int slot) {
 
 	@Override
 	protected void drawGuiContainerBackgroundLayer(float f, int x, int y) {
-		GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-		mc.renderEngine.bindTexture(_chassiPipe.getChassiGUITexture());
-		drawTexturedModalRect(guiLeft, guiTop, 0, 0, xSize, ySize);
+		GuiGraphics.drawGuiBackGround(mc, guiLeft, guiTop, right, bottom, zLevel, true);
+
+		for(int  i=0; i<_chassiPipe.getChassiSize(); i++)
```
Comment (bziemons) [inline common/logisticspipes/gui/GuiChassiPipe.java]: styling

File: common/logisticspipes/network/guis/pipe/ChassiGuiProvider.java [hunk 3]
```diff
@@ -46,23 +46,9 @@ public DummyContainer getContainer(EntityPlayer player) {
...
-		if (_chassiPipe.getChassiSize() > 3) {
-			dummy.addModuleSlot(3, _moduleInventory, 19, 69, _chassiPipe);
-		}
-		if (_chassiPipe.getChassiSize() > 4) {
-			dummy.addModuleSlot(4, _moduleInventory, 19, 89, _chassiPipe);
-			dummy.addModuleSlot(5, _moduleInventory, 19, 109, _chassiPipe);
-			dummy.addModuleSlot(6, _moduleInventory, 19, 129, _chassiPipe);
-			dummy.addModuleSlot(7, _moduleInventory, 19, 149, _chassiPipe);
+
+		for(int i=0; i<_chassiPipe.getChassiSize(); i++){
```
Comment (bziemons) [inline common/logisticspipes/network/guis/pipe/ChassiGuiProvider.java]: styling

File: common/logisticspipes/config/Configs.java [hunk 4]
```diff
@@ -39,6 +39,10 @@
 	public static boolean TOOLTIP_INFO = LPConstants.DEBUG;
 	public static boolean ENABLE_PARTICLE_FX = true;
 
+
+	public static int [] CHASSI_SLOTS_ARRAY = {1,2,3,4,8};
```
Comment (bziemons) [inline common/logisticspipes/config/Configs.java:42]: please only one new-line before and after this field, styling

File: common/logisticspipes/config/Configs.java [hunk 5]
```diff
@@ -202,6 +206,14 @@ public static void load() {
 						"Do you fancy easter eggs?")
 						.getBoolean(false);
 
+
+		Configs.CHASSI_SLOTS_ARRAY = Configs.CONFIGURATION
```
Comment (bziemons) [inline common/logisticspipes/config/Configs.java:207]: again only one new-line before and after this

File: common/logisticspipes/gui/GuiChassiPipe.java [hunk 6]
```diff
@@ -61,29 +58,11 @@ public GuiChassiPipe(EntityPlayer player, PipeLogisticsChassi chassi, boolean ha
...
-		if (_chassiPipe.getChassiSize() > 4) {
-			dummy.addModuleSlot(4, _moduleInventory, 19, 89, _chassiPipe);
-			dummy.addModuleSlot(5, _moduleInventory, 19, 109, _chassiPipe);
-			dummy.addModuleSlot(6, _moduleInventory, 19, 129, _chassiPipe);
-			dummy.addModuleSlot(7, _moduleInventory, 19, 149, _chassiPipe);
-		}
+
+		dummy.addNormalSlotsForPlayerInventory(18, 9 + 20* _chassiPipe.getChassiSize());
+
+		for(int  i=0; i<_chassiPipe.getChassiSize(); i++)
```
Comment (bziemons) [inline common/logisticspipes/gui/GuiChassiPipe.java]: styling

File: common/logisticspipes/gui/GuiChassiPipe.java [hunk 7]
```diff
@@ -170,24 +149,9 @@ protected void drawGuiContainerForegroundLayer(int par1, int par2) {
...
-			mc.fontRenderer.drawString(getModuleName(3), 40, 74, 0x404040);
-		}
-		if (_chassiPipe.getChassiSize() > 4) {
-			mc.fontRenderer.drawString(getModuleName(4), 40, 94, 0x404040);
-			mc.fontRenderer.drawString(getModuleName(5), 40, 114, 0x404040);
-			mc.fontRenderer.drawString(getModuleName(6), 40, 134, 0x404040);
-			mc.fontRenderer.drawString(getModuleName(7), 40, 154, 0x404040);
-		}
+
+		for(int  i=0; i<_chassiPipe.getChassiSize() ; i++)
```
Comment (bziemons) [inline common/logisticspipes/gui/GuiChassiPipe.java]: styling

**Review by bziemons — CHANGES_REQUESTED** [review]
If I change the config option to only have 4 values: `[1,2,3,4]` this code would crash with an IndexOutOfBoundsException in the Mk5 Pipe. Please ensure the config option has always 5 entries and add an appropriate hint to the config option's description.

**Comment by Noradly** [comment]
Added your check and a check to make certain that the values are >=0. The issues above (should) have been resolved.

**Comment by bziemons** [comment]
With styling I mean the usage of the formatter. The IntelliJ project has the formatter pre-configured and you can just run it.

Otherwise, what I meant was:
```diff
- for(int  i=0; i<_chassiPipe.getChassiSize() ; i++){
+ for (int i = 0; i < _chassiPipe.getChassiSize(); i++) {
```

Please just stick with the project's existing code style.

**Comment by Noradly** [comment]
Thanks, added an extra test to make certain that the entries are in ascending order (we use the last value to set the size of Slot[] upgradeslot in common/logisticspipes/gui/GuiChassiPipe.java).

**Comment by Noradly** [comment]
Removed the first initialization of upgradeConfig since it was initialized twice.

**Comment by Noradly** [comment]
I have found an issue though, the tooltip of a chassipipe says how many modules a pipe can hold, this text comes from the en_us.lang file and should be dependent on the actual number of modules a pipe can hold.

File: common/logisticspipes/config/Configs.java [hunk 8]
```diff
@@ -202,6 +204,32 @@ public static void load() {
...
+						"The number of slots in a chassi pipe starting from MK1 to MK5, because there are 5 tiers there need to be 5 values(positive integers, zero is allowed) in ascending order.")
+						.getIntList();
+
+		if (Configs.CHASSI_SLOTS_ARRAY.length != 5) {
+			throw new RuntimeException(
+					"The config file of Logistics Pipes needs to have 5 values(positive integers, zero is allowed) in ascending order in chassiSlots. \nYou have given us: "
+							+ Configs.CHASSI_SLOTS_ARRAY.length + " values.");
+		}
+
+		for(int i=0; i<Configs.CHASSI_SLOTS_ARRAY.length; i++){
```
Comment (bziemons) [inline common/logisticspipes/config/Configs.java]: styling here as well, please

File: common/logisticspipes/config/Configs.java [hunk 9]
```diff
@@ -202,6 +204,32 @@ public static void load() {
...
+			if(Configs.CHASSI_SLOTS_ARRAY[i]<0)
+				throw new RuntimeException(
+					"The config file of Logistics Pipes needs to have 5 values(positive integers, zero is allowed) in ascending order in chassiSlots. \nYou have given us: "
+							+ Configs.CHASSI_SLOTS_ARRAY[i] + " at index: " + i);
+		}
+
+		for(int i=0; i<Configs.CHASSI_SLOTS_ARRAY.length-1; i++){
```
Comment (bziemons) [inline common/logisticspipes/config/Configs.java]: styling here as well, please

**Comment by bziemons** [comment]
> I have found an issue though, the tooltip of a chassipipe says how many modules a pipe can hold, this text comes from the en_us.lang file and should be dependent on the actual number of modules a pipe can hold.

Add "by default" to the english language for now. If you like, you can add placeholders for language strings in another PR.

**Review by bziemons — APPROVED** [review]

**Comment by bziemons** [comment]
Someone needs to test the changes and then this is good to merge :) Thanks for the work @Noradly

**Comment by Noradly** [comment]
@theZorro266 I have been playing with this branch for a couple days now (since 9d4d41d) and all the bugs I've found have been fixed.

**Comment by davboecki** [comment]
These changes don't work. While it appears to work for the GUI changes. As soon as you save and reload the world, the modules are removed except for the inv slots that are supposed to be there by default.

**Comment by bziemons** [comment]
@davboecki that behavior actually happens with or without these changes. Just tested that.

I squashed and merged this code locally. I will try to find a fix for the mentioned problem.

**Comment by bziemons** [comment]
Also, I am very sorry for the misunderstanding. This should have been solved with a bit more communicating and testing, I think.

**Pull request closed without merging**